### PR TITLE
task done

### PR DESCRIPTION
**Pull request by AhmedMohmd1**

### CODE REVIEW

**Comment by bego00** [comment]
Test case 9: Search product 
Case follows:
1. Launch browser
2. Navigate to URL ['http://automationexercise.com'](https://automationexercise.com/)
3. Verify that the home page is visible successfully
4. Click on the 'Products' button
5. Verify the user is navigated to the ALL PRODUCTS page successfully
6. Enter the product name in the search input and click the search button
7. Verify 'SEARCHED PRODUCTS' is visible
8. Verify all the products related to the search are visible

File: pom.xml [hunk 1]
```diff
@@ -19,20 +19,36 @@
     <build>
         <directory>test-outputs/target</directory>
     </build>
+
     <dependencies>
         <!-- https://mvnrepository.com/artifact/org.seleniumhq.selenium/selenium-java -->
         <dependency>
             <groupId>org.seleniumhq.selenium</groupId>
             <artifactId>selenium-java</artifactId>
-            <version>${seleniumVersion}</version>
```
Comment (Ashraaf7) [inline pom.xml:27]: Why did you delete the ${seleniumVersion}? I have created it and it binds to the variable on the properties tag for easy handling of the versioning.

File: pom.xml [hunk 2]
```diff
@@ -19,20 +19,36 @@
...
+            <version>4.18.1</version>
         </dependency>
 
+
         <!-- https://mvnrepository.com/artifact/org.testng/testng -->
         <dependency>
             <groupId>org.testng</groupId>
             <artifactId>testng</artifactId>
-            <version>${testNGVersion}</version>
```
Comment (Ashraaf7) [inline pom.xml:34]: Same here.

File: pom.xml [hunk 3]
```diff
@@ -19,20 +19,36 @@
...
+            <version>7.9.0</version>
             <scope>test</scope>
         </dependency>
+
+        <!-- https://mvnrepository.com/artifact/io.github.bonigarcia/webdrivermanager -->
+        <dependency>
+            <groupId>io.github.bonigarcia</groupId>
+            <artifactId>webdrivermanager</artifactId>
+            <version>5.7.0</version>
```
Comment (Ashraaf7) [inline pom.xml:44]: you did not need a web driver manager.

File: pom.xml [hunk 4]
```diff
@@ -19,20 +19,36 @@
...
+        </dependency>
+        <!-- https://mvnrepository.com/artifact/org.seleniumhq.selenium/selenium-manager -->
+        <dependency>
+            <groupId>org.seleniumhq.selenium</groupId>
+            <artifactId>selenium-manager</artifactId>
```
Comment (Ashraaf7) [inline pom.xml:49]: duplicated

File: src/main/java/com/automationexercise/Pages/BasePage.java
Comment (Ashraaf7) [inline src/main/java/com/automationexercise/Pages/BasePage.java:1]: All these functions will be on Utils classes so you can delete the BasePage and hold until it is created.

File: src/main/java/com/automationexercise/Pages/HomePage.java [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+public class HomePage extends BasePage {
+
+
+    public HomePage(WebDriver driver) {
+        super(driver);
+    }
+
+    By ProductButton = By.xpath("//a[@href=\"/products\"]");
+    WebElement AllProductsPage = driver.findElement
```
Comment (Ashraaf7) [inline src/main/java/com/automationexercise/Pages/HomePage.java:18]: You should save any WebElement object because it will cause an StaleElementRefrence. Just you save the locator as By object and find it whenever you need.
Check this:https://www.selenium.dev/documentation/webdriver/troubleshooting/errors/#stale-element-reference-exception

File: src/test/java/Base/BaseTest.java
Comment (Ashraaf7) [inline src/test/java/Base/BaseTest.java:1]: Kindly, hold until creating the drive manager class so you can use it here.

File: src/test/java/Base/BaseTest.java [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+import org.testng.annotations.AfterTest;
+import org.testng.annotations.BeforeTest;
+import org.testng.asserts.SoftAssert;
+
+import java.io.File;
+import java.io.IOException;
+
+public class BaseTest {
+
+    public static final String BASE_URL = "https://automationexercise.com/";
```
Comment (Ashraaf7) [inline src/test/java/Base/BaseTest.java:22]: We will read all data from external files

File: src/test/java/Base/BaseTest.java [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+    public WebDriver driver;
+    protected SoftAssert soft;
+    public HomePage homePage;
+
+    @BeforeTest
```
Comment (Ashraaf7) [inline src/test/java/Base/BaseTest.java:27]: Should be before and after **Method**

File: src/test/java/com/automationexercise/tests/Tests/HomePageTests.java [hunk 8]
```diff
@@ -0,0 +1,19 @@
+package com.automationexercise.tests.Tests;
+
+import Base.BaseTest;
+import org.testng.annotations.Test;
+
+public class HomePageTests extends BaseTest {
+
+    @Test
+    public void test() {
```
Comment (Ashraaf7) [inline src/test/java/com/automationexercise/tests/Tests/HomePageTests.java:9]: Each test case should be named with a readable name.

File: src/test/java/com/automationexercise/tests/Tests/HomePageTests.java [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+        soft.assertEquals(homePage.VerifyHomePageVisible(), "https://automationexercise.com/", "You are not in HOMEPAGE");
+        homePage.PressProductsButton();
+        soft.assertTrue(homePage.VerifyUserNavigatedToAllProductsPage());
+        homePage.EnterProduct("Tshirt");
+        homePage.PressSearchButton();
+        soft.assertTrue(homePage.VerifySearchedProductsSectionVisible());
+        soft.assertEquals(homePage.VerifyAllProductRelatedToSearchVisible(), 6);
```
Comment (Ashraaf7) [inline src/test/java/com/automationexercise/tests/Tests/HomePageTests.java:16]: You forgot assertAll();

**Review by Ashraaf7 — CHANGES_REQUESTED** [review]

File: pom.xml (hunk 3 above)
Comment (bego00) [reply pom.xml:44]: I use it in the BaseTest Class as shown in the screenshot
Screenshot 2024-03-17 132639

Comment (Ashraaf7) [reply pom.xml:44]: Check here:https://bonigarcia.dev/webdrivermanager/#webdrivermanager-and-selenium-manager
image
@bego00
